### PR TITLE
Introduces *_extensions to extend functionality in hamilton

### PR DESCRIPTION
**Pull request by skrawcz**
Introduces *_extensions to extend functionality in hamilton

So we had a problem in the decorators (notably extract_columns)
that assumed pandas. We need to abstract this out so that other
dataframe libraries could be easily used with the decorator.

This is a commit that introduces a few changes to do that:

(1) We decided on a plugins module that houses two types of files:
*_extensions.py, and *_implementations.py. The *_extensions includes
code that the decorators and hamilton framework depends on.
The *_implementations will house specific implementations related to
that dataframe library for example, e.g. result builders. Users will reference these.
 Note we haven't created any _implementations.py files in this PR -- will do so in later ones.

(2) We decide that there are several functions that need to live in _extensions.py
that enable the hamilton framework to delegate/find the right thing to use.
This could be an object -- that one has to implement. But because it's already
in the module -- and we want things to be registered, I think this abstraction right
now is simpler -- the housing unit is the module that the functions are in -- don't need
to put them into an object right now.

(3) I decided trying to import the modules and having them determine how to error
would be the best way to determine if the rest of the module import should proceed.

## Changes
 - adds plugins module
 - adds to hamilton.base.py  to house registering extension options
 - adds extensions for polars, pandas, and pandas on pyspark
 - adds polars example and tweaks pandas on pyspark one to exercise extract_columns

## How I tested this
Ran code locally.

## Notes
This will help solve a user issue.

## Checklist

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future TODOs are captured in comments
- [x] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

File: hamilton/base.py [hunk 1]
```diff
@@ -21,6 +23,47 @@
...
+        if series_type is not None:
+            return series_type
+    raise NotImplementedError(f"Unknown dataframe type: {dataframe_type}. Cannot get Column type.")
+
+
+plugins_modules = ["pandas", "polars", "pyspark"]
+
+for plugin_module in plugins_modules:
+    try:
+        importlib.import_module(f"hamilton.plugins.{plugin_module}_extensions")
```
Comment (elijahbenizzy) [inline hamilton/base.py:62]: check for functions here to ensure "interface"/

Comment (elijahbenizzy) [reply hamilton/base.py:62]: and the logger.info("detected X support, loading extension for it")

Comment (elijahbenizzy) [reply hamilton/base.py:62]: logger.debug("did not detect...")

File: hamilton/base.py [hunk 2]
```diff
@@ -20,6 +22,81 @@
...
+    mod = importlib.import_module(f"hamilton.plugins.{plugin_module}_extensions")
+    assert hasattr(mod, "DATAFRAME_TYPE"), "Error extension mission DATAFRAME_TYPE"
+    assert hasattr(mod, "COLUMN_TYPE"), "Error extension mission COLUMN_TYPE"
+    assert hasattr(
+        mod, f"get_column_{plugin_module}"
+    ), f"Error extension mission get_column_{plugin_module}"
+    assert hasattr(
+        mod, f"fill_with_scalar_{plugin_module}"
+    ), f"Error extension mission fill_with_scalar_{plugin_module}"
+    global DF_TYPE_AND_COLUMN_TYPES
```
Comment (elijahbenizzy) [inline hamilton/base.py:84]: So I think this works -- only weird thing is that we do this differently than we do registry for single dispatch.

- Modules are responsible for registering their own implementations
- Types are added centrally...

Might be nice to have modules add their own types?

File: hamilton/base.py [hunk 3]
```diff
@@ -20,6 +22,81 @@
...
+        "dataframe_type": mod.DATAFRAME_TYPE,
+        "column_type": mod.COLUMN_TYPE,
+    }
+    logger.info(f"Detected {plugin_module} and successfully loaded Hamilton extensions.")
+
+
+plugins_modules = ["pandas", "polars", "pyspark_pandas"]
+for plugin_module in plugins_modules:
+    try:
+        _load_extension(plugin_module)
```
Comment (elijahbenizzy) [inline hamilton/base.py:95]: Isn't this a circular dependency?
- plugin_module depends on base to register
- we import plugin_module here
- this happens at import time...

File: hamilton/function_modifiers/expanders.py [hunk 4]
```diff
@@ -335,9 +335,12 @@ def validate(self, fn: Callable):
...
         """
         output_type = inspect.signature(fn).return_annotation
-        if not issubclass(output_type, pd.DataFrame):
+        try:
+            hamilton_base.get_column_type_from_df_type(output_type)
+        except NotImplementedError:
             raise base.InvalidDecoratorException(
-                f"For extracting columns, output type must be pandas dataframe, not: {output_type}"
+                # TODO: capture was dataframe libraries are supported and print here.
+                f"Error {fn} does not output a type we know about. Is it a dataframe type we support?"
```
Comment (elijahbenizzy) [inline hamilton/function_modifiers/expanders.py:343]: Maybe a link to open an issue/reach out?

File: hamilton/plugins/pandas_extensions.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+try:
+    import pandas as pd
+except ImportError:
+    raise NotImplementedError("Pandas is not installed.")
+
+from hamilton import base
+
+# Required constants!
+DATAFRAME_TYPE = pd.DataFrame
```
Comment (elijahbenizzy) [inline hamilton/plugins/pandas_extensions.py:11]: If we want we can statically validate these extensions to make sure we don't change anything -- parse an AST and look for these nodes...

File: hamilton/plugins/polars_implementations.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
+from typing import Dict
+
+import polars as pl
+
+from hamilton import base
+
+
+class PolarsDataFrameResult(base.ResultMixin):
```
Comment (elijahbenizzy) [inline hamilton/plugins/polars_implementations.py:8]: Thought -- if we had a general "concat" operation or "join" operation we could implement a more generic result builder. We now have the ability to register operations...

**Review by elijahbenizzy — COMMENTED** [review]
LGTM -- some qs about circular dependencies but I imagine you've tested it and I'm missing something...

File: hamilton/base.py (hunk 2 above)
Comment (skrawcz) [reply hamilton/base.py:84]: yep

File: hamilton/base.py (hunk 3 above)
Comment (skrawcz) [reply hamilton/base.py:95]: yep

File: hamilton/function_modifiers/expanders.py (hunk 4 above)
Comment (skrawcz) [reply hamilton/function_modifiers/expanders.py:342]: They should get the pointer to slack when this occurs.

File: hamilton/plugins/polars_implementations.py (hunk 6 above)
Comment (skrawcz) [reply hamilton/plugins/polars_implementations.py:8]: Maybe. But polars doesn't have an index -- so 🤷.

**Pull request merged**